commit/chainfee/processor.go

```diff
@@ -3,41 +3,137 @@ package chainfee
+func (p *processor) Query(ctx context.Context, prevOutcome Outcome) (Query, error) {
 	return Query{}, nil
 }
 
-func (w *Processor) Observation(
+func (p *processor) Observation(
 	ctx context.Context,
 	prevOutcome Outcome,
 	query Query,
 ) (Observation, error) {
-	return Observation{}, nil
+	// Get the fee components for all available chains that we can read from
+	feeComponents := p.ccipReader.GetAvailableChainsFeeComponents(ctx)
+	// Get the native token prices for all available chains that we can read from
+	nativeTokenPrices := p.ccipReader.GetWrappedNativeTokenPriceUSD(ctx, maps.Keys(feeComponents))
+	// Get the latest chain fee price updates for the source chains
+	timestampedPriceUpdates := p.ccipReader.GetChainFeePriceUpdate(ctx, maps.Keys(feeComponents))
+
+	return Observation{
+		FChain:            p.ObserveFChain(),
+		FeeComponents:     feeComponents,
+		NativeTokenPrices: nativeTokenPrices,
+		ChainFeeUpdates:   FeeUpdatesFromTimestampedBig(timestampedPriceUpdates),
+		TimestampNow:      time.Now().UTC(),
+	}, nil
 }
 
-func (w *Processor) Outcome(
+func (p *processor) Outcome(
 	prevOutcome Outcome,
 	query Query,
 	aos []plugincommon.AttributedObservation[Observation],
 ) (Outcome, error) {
-	return Outcome{}, nil
+
+	consensusObs, err := p.getConsensusObservation(aos)
+	if err != nil {
+		return Outcome{}, fmt.Errorf("failed to get consensus observation: %w", err)
+	}
+	// No need to update yet
+	if len(consensusObs.FeeComponents) == 0 {
+		p.lggr.Debug("no consensus on fee components, nothing to update",
+			"consensusObs", consensusObs)
+		return Outcome{}, nil
+	}
+
+	// Stop early if earliest updated timestamp is still fresh
+	//earliestUpdateTime := consensus.EarliestTimestamp(maps.Values(consensusObs.ChainFeeUpdates))
+	//nextUpdateTime := earliestUpdateTime.Add(p.ChainFeePriceBatchWriteFrequency.Duration())
+	//if nextUpdateTime.After(consensusObs.TimestampNow) {
+	//	return Outcome{}, nil
+	//}
+
+	chainFeeUSDPrices := make(map[cciptypes.ChainSelector]ComponentsUSDPrices)
+	// We need to report a packed GasPrice
+	// The packed GasPrice is a 224-bit integer with the following format:
+	// (dataAvFeePriceUSD) << 112 | (executionFeePriceUSD)
+	// nolint:lll
+	// https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/FeeQuoter.sol#L498
+	// In next loop we calculate the price in USD for the data availability and execution fees.
+	// And getGasPricesToUpdate will select and calculate the **packed** gas price to update based.
+	for chain, feeComp := range consensusObs.FeeComponents {
+		nativeTokenPriceUSD := consensusObs.NativeTokenPrices[chain].Int
+
+		// Calculate the price in USD for the data availability and execution fees.
+		// Raw fee components are in native token units
+		chainFeeUsd := ComponentsUSDPrices{
+			ExecutionFeePriceUSD: new(big.Int).Mul(nativeTokenPriceUSD, feeComp.ExecutionFee),
```

Comment:
So this is native token price in USD to 36 decimals, ExecutionFee is the fee in the native token its the smallest denomination (e.g if its ethereum, its wei)?

So this means `USD/NATIVE * NATIVE == USD`, but the question I have is, is `nativeTokenPriceUSD` per smallest denomination, or is it per 1e18 of the smallest denomination? The denominations have to be the same in order for this multiplication to make sense.